### PR TITLE
fixes #396 - do not trigger IDE internal error when user selects a cloud project to debug to which he has no access

### LINKED ISSUE #396: IDE internal error when selecting a forbidden project
When selecting a project to debug in the debug dialog for which the user does not have permission (an HTTP 403 is returned), the IDE reports an an "Internal IDE error has occurred":

image

This should probably not be reported as an internal IDE error, and just display the error message in the dialog as usual.

### PR DESCRIPTION
**Pull request by akerekes**

### CODE REVIEW

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java [hunk 1]
```diff
@@ -199,7 +199,7 @@ public void run() {
             }
           });
 
-          LOG.error("Error listing debuggees from Cloud Debugger API", ex);
+          LOG.warn("Error listing debuggees from Cloud Debugger API", ex);
```
Comment (loosebazooka) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: do we need to tell the user in other ways as well?

Comment (akerekes) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: The user sees the message to select another project with debuggable module in the dialog.

Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: What's the exact message?  To fix #396 we need to let the user know the
difference between a project with no debuggable modules, and a project for
which they don't have debug access.

Comment (akerekes) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: work in progress to give different message based on the reason for not being able to select modules

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java [hunk 2]
```diff
@@ -154,7 +155,7 @@ public void run() {
                 DebugTarget targetSelection = null;
 
                 if (debuggees == null || debuggees.getDebuggees() == null || debuggees.getDebuggees().isEmpty()) {
-                  disableTargetSelector();
+                  disableTargetSelector(new NoDebuggableModuleException());
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: @elharo what do you think of this?  It feels kind of wrong to me to create an exception type just to be able to reuse a throwable based API.

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/AbstractDebugTarget.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.google.gct.idea.debugger.ui;
+
+/**
+ * Parent class to serve as a base for actual debug targets or specific child classes used e.g. for
+ * error handling
+ */
+public abstract class AbstractDebugTarget {
```
Comment (patflynn) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/AbstractDebugTarget.java]: is this a marker type?  I don't see anywhere that's checking it.

Comment (akerekes) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/AbstractDebugTarget.java]: The idea was here to use a consistent base type for the combobox used to display debug targets. Unfortunately Java supports generic JComboBox only from JDK7 and up, so we can happily mix different types as items without the compiler complaining.

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java (hunk 2 above)
Comment (akerekes) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: In a similar change I've decided to just create 2 methods, one with a String parameter and the other delegating to it after figuring out the message. I also don't feel strongly about this usage of Exception.

File: google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties [hunk 4]
```diff
@@ -129,4 +129,6 @@ clouddebug.couldnotconnect=Could not connect to Google. Please make sure you hav
 clouddebug.stop.and.create.new.session=A cloud debugger session is already running.\nWould you like to stop it and start a new session?
 clouddebug.stop.background.and.create.new.session=A cloud debugger session is listening in the background.\nWould you like to stop it and start a new session?
 clouddebug.background.listener.error.title=Error while connecting to Cloud Debugger backend
-clouddebug.background.listener.error.message=A cloud debugger process is listening in the background but cannot access the backend. Verify that there is a Google account signed in with access to the project.
\ No newline at end of file
+clouddebug.background.listener.error.message=A cloud debugger process is listening in the background but cannot access the backend. Verify that there is a Google account signed in with access to the project.
+clouddebug.debug.targets.error=Error accessing debug targets: {0}
+clouddebug.debug.targets.accessdenied=Access denied, you may not have the necessary permissions to run Cloud Debugger with this project
```
Comment (elharo) [inline google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties]: denied. You may not
also, period at end

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/DebugTarget.java [hunk 5]
```diff
@@ -29,7 +29,7 @@
 /**
  * The class models out the details for a target debuggable module.
  */
-class DebugTarget {
+class DebugTarget extends AbstractDebugTarget {
```
Comment (elharo) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/DebugTarget.java]: Could this be named something more specific, and AbstractDebugTarget just be DebugTarget?

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java [hunk 6]
```diff
@@ -191,27 +192,48 @@ public void run() {
...
+    if (reason instanceof NoDebuggableModuleException) {
+      return GctBundle.getString("clouddebug.selectvalidproject");
+    } else if (reason instanceof GoogleJsonResponseException) {
+      return resolveJsonResponseToMessage((GoogleJsonResponseException) reason);
+    } else {
+      return GctBundle.getString("clouddebug.debug.targets.error", reason.getLocalizedMessage());
+    }
+  }
+
+  private String resolveJsonResponseToMessage(GoogleJsonResponseException reason) {
```
Comment (elharo) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: perhaps static

File: google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java [hunk 7]
```diff
@@ -191,27 +192,48 @@ public void run() {
...
-      targetSelector.addItem(moduleWarning);
+    String moduleWarning = resolveErrorToMessage(reason);
+    if(targetSelector.getSelectedItem() instanceof ErrorDebugTarget) {
+      ((ErrorDebugTarget) targetSelector.getSelectedItem()).setErrorMessage(moduleWarning);
+    } else {
+      targetSelector.addItem(new ErrorDebugTarget(moduleWarning));
+    }
+  }
+
+  private String resolveErrorToMessage(Throwable reason) {
```
Comment (elharo) [inline google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ui/ProjectDebuggeeBinding.java]: perhaps static

**Comment by elharo** [comment]
lgtm

**Comment by patflynn** [comment]
@akerekes what's the status of this PR?

**Comment by akerekes** [comment]
@patflynn I've addressed @elharo's comments, ready for review. Once I merge #395 to master, I'll update this with master and it's ready for final review.

**Pull request merged**